contracts/near/eth-prover/src/lib.rs

```diff
@@ -204,112 +208,139 @@ impl EthProver {
         let dec = Rlp::new(&node.as_slice());
 
         if key_index == 0 {
-            // trie root is always a hash
-            assert_eq!(near_keccak256(node), (expected_root.0).0);
+            if near_keccak256(node) != (expected_root.0).0 {
+                return false;
+            }
         } else if node.len() < 32 {
             // if rlp < 32 bytes, then it is not hashed
-            assert_eq!(dec.as_raw(), (expected_root.0).0);
+            if dec.as_raw() != (expected_root.0).0 {
+                return false;
+            }
         } else {
-            assert_eq!(near_keccak256(node), (expected_root.0).0);
+            if near_keccak256(node) != (expected_root.0).0 {
+                return false;
+            }
         }
 
-        if dec.iter().count() == 17 {
-            // branch node
+        let item_count = if let Ok(item_count) = dec.item_count() {
+            item_count
+        } else {
+            return false;
+        };
+
+        if item_count == 17 {
+            // Branch node
             if key_index == key.len() {
-                if dec
-                    .at(dec.iter().count() - 1)
-                    .unwrap()
-                    .as_val::<Vec<u8>>()
-                    .unwrap()
-                    == expected_value
-                {
-                    // value stored in the branch
-                    return true;
-                }
+                // The key was fully traversed, so we should check the value stored at this node
+                dec.at(item_count - 1) // The value is in the last field of the branch node
+                    .and_then(|rlp| rlp.as_val::<Vec<u8>>())
+                    .map(|value| value == expected_value)
+                    .unwrap_or_default()
             } else if key_index < key.len() {
```

Comment:
We can probably assume that `key_index <= key.len()`, so this check is not necessary.